homeassistant/components/emulated_kasa/__init__.py

```diff
@@ -0,0 +1,123 @@
+    conf = config.get(DOMAIN)
+    if not conf:
+        return True
+    entity_configs = conf[CONF_ENTITIES]
+
+    def devices():
+        """Devices to be emulated."""
+        yield from get_plug_devices(hass, entity_configs)
+
+    server = SenseLink(devices)
+
+    async def stop_emulated_kasa(event):
+        await server.stop()
+
+    async def start_emulated_kasa(event):
+        validate_configs(hass, entity_configs)
+        try:
+            await server.start()
+        except OSError as error:
+            _LOGGER.error("Failed to create UDP server at port 9999: %s", error)
+        else:
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_emulated_kasa)
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, start_emulated_kasa)
+
+    return True
+
+
+def validate_configs(hass, entity_configs):
+    """Validate that entities exist and ensure templates are ready to use."""
+    for entity_id, entity_config in entity_configs.items():
+        state = hass.states.get(entity_id)
+        if state is None:
+            _LOGGER.debug("Entity not found: %s", entity_id)
+            continue
+        if CONF_POWER in entity_config:
+            power_val = entity_config[CONF_POWER]
+            if isinstance(power_val, str) and is_template_string(power_val):
+                entity_config[CONF_POWER] = Template(power_val, hass)
+            elif isinstance(power_val, Template):
+                entity_config[CONF_POWER].hass = hass
+        elif state.domain == SENSOR_DOMAIN:
+            continue
+        elif ATTR_CURRENT_POWER_W in state.attributes:
+            continue
+        else:
+            _LOGGER.debug("No power value defined for: %s", entity_id)
+
+
+def get_plug_devices(hass, entity_configs):
+    """Produce list of plug devices from config entities."""
+    for entity_id, entity_config in entity_configs.items():
+        state = hass.states.get(entity_id)
+        if state is None:
+            continue
+        name = entity_config.get(CONF_NAME, state.name)
+
+        if state.state == STATE_ON or state.domain == SENSOR_DOMAIN:
+            if CONF_POWER in entity_config:
+                power_val = entity_config[CONF_POWER]
+                if isinstance(power_val, (float, int)):
+                    power = float(power_val)
+                elif isinstance(power_val, str):
+                    power = float(hass.states.get(power_val).state)
+                elif isinstance(power_val, Template):
+                    power = float(power_val.async_render())
+            elif ATTR_CURRENT_POWER_W in state.attributes:
+                power = float(state.attributes[ATTR_CURRENT_POWER_W])
+            elif state.domain == SENSOR_DOMAIN:
+                power = float(state.state)
+        else:
+            power = 0.0
+        last_changed = state.last_changed.timestamp()
+        yield PlugInstance(entity_id, start_time=last_changed, alias=name, power=power)
```

Comment:
I'm assuming entity_id is being used to unique identifier here?

If the entity has a globally unique id it may make sense to pass that instead so they don't loose their power history when the entity gets renamed.

Example from homekit.aidmanager

def get_system_unique_id(entity: RegistryEntry):
    """Determine the system wide unique_id for an entity."""
    return f"{entity.platform}.{entity.domain}.{entity.unique_id}"